### PR TITLE
[Performance] Avoid 500MB of HashSet enumerator allocations calling `Any`

### PR DESCRIPTION
**Pull request by Youssef1313**
image

The new extension method is now picked instead of LINQ Any.

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Can you include the after trace

File: src/Compilers/Core/Portable/Collections/HashSetExtensions.cs [hunk 1]
```diff
@@ -28,5 +29,18 @@ internal static bool InitializeAndAdd<T>([NotNullIfNotNull(parameterName: nameof
 
             return hashSet.Add(item);
         }
+
+        internal static bool Any<T>(this HashSet<T> hashSet, Func<T, bool> predicate)
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/Collections/HashSetExtensions.cs:33]: Can we include in the comment why this was introduced so future developers don't delete it?

**Comment by tmat** [comment]
Should this be addressed in the runtime libraries? Perhaps the implementation of Enumerable.Any should get specialized for common collection types that have struct enumerator?

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 2)

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler For the second review.

**Comment by CyrusNajmabadi** [comment]
> Should this be addressed in the runtime libraries? Perhaps the implementation of Enumerable.Any should get specialized for common collection types that have struct enumerator?

tagging @stephentoub My understanding is that it's a tradeoff as now there are more runtime checks that have to happen.  Perhaps there might be a quick internal interface check that could say "i support no-alloc enumeration", and if that is true, then a call to a helper which checks the types?

**Review by cston — APPROVED** [review]

**Comment by AlekseyTs** [comment]
@Youssef1313 Thank you for the contribution

**Pull request merged**